docs/schulcloud-server/Coding-Guidelines/testing.md

```diff
@@ -0,0 +1,376 @@
+
+- what component is being tested
+- under what condition
+- the expected outcome
+
+To facilitate this, your tests should be wrapped in at least two describe levels.
+
+```TypeScript
+// Name of the unit under test
+describe("Course Service", (() => {
+    // method that is called
+    describe('createCourse', () => {
+        // a "when..." sentence
+        describe("When a student tries to create a course", (() => {
+            // a "should..." sentence
+                it("should return course", async () => {
+                    ...
+                });
+        });
+    });
+});
+```
+
+### Isolation
+
+Each test should be able to run alone, as well as together with any other tests. To ensure this, it is important that the test does not depend on any preexisting data.
+
+- Each test should generate the data it needs, and ensure that its data is deleted afterwards. (this is usually done via mocha's "afterEach" function.
+- When you create objects with fields that have to be globally unique, like the account username, you must ensure the name you choose is unique. This can be done by including a timestamp.
+- Never use seeddata.
+
+### Test Structure
+
+Your test should be structured in three seperate areas, each distinguished by at least an empty line:
+
+- Arrange - set up your testdata
+- Act - call the function you want to test
+- Assert - check the result
+
+this is known as the AAA-pattern.
+
+The tests for a unit should cover as much scenarios as possible. Parameters and the combination of parameters can often take numerous values. Therefore it largely differs from case to case what a sufficient amount of scenarios would be. Parameter values that contradict the typescript type definition should be ignored as a test case.
+The test coverage report already enforces scenarios that test every possible if/else result in the code. But still some scenarios are not covered by the report and must be tested:
+
+- All error scenarios: That means one describe block for every call that can reject.
+
+We use different levels of describe blocks to structure the tests in a way, that the tested scenarios could easily be recognized. The outer describe would be the function call itself. Every scenario is added as another describe inside the outer describe.
+
+All of the data and mock preparation should happen in a setup function. Every describe scenario only contains one setup function and is called in every test. No further data or mock preparation should be added to the test. Often there will be only one test in every describe scenario, this is perfectly fine with our desired structure.
+
+```TypeScript
+describe('[method]', () => {
+    describe('when [senario description that is prepared in setup]', () => {
+        const setup = () => {
+            // prepare the data and mocks for this scenario
```

Comment:
I miss the part of deep mocked services should be mocked. Do not use Jest (expect it is a other method of the tested service in this file), use Once, use for async Resolve / Reject for syncron mockImplementation.